Add compartment_set stimuli support

ilkilic opened the pull request:
## Summary

This PR adds support for morphology-derived stimulus targeting through `MorphologyLocations` and SONATA `compartment_set` materialization.

Implemented:
- `MorphologyLocations` blocks in simulation configs
- stimulus targeting via generated compartment sets
- morphology-location materialization pipeline
- dedicated `Targeting` UI group
- backward compatibility for legacy `*SomaticStimulus` aliases
- integration with the new typed neuron-set references

## What is included

### Morphology targeting

Stimuli can now target:

- a neuron set
- an explicit `CompartmentSet`
- generated `MorphologyLocations`

`MorphologyLocations` uses the new typed biophysical neuron-set references to select the cells on which locations are generated.

### Materialization pipeline
- Morphology locations are materialized into SONATA `compartment_set` entries during simulation generation
- Generated compartment sets are exported into `compartment_sets.json`
- Multiple generated compartment sets are consolidated into a single SONATA file

### UI/schema
- Added dedicated `Targeting` section
- Added `MorphologyLocations` references to supported stimuli
- `section_types` support exists in the backend but is currently hidden from the generated UI pending follow-up work

### Compatibility
- Added deserialization support for legacy `*SomaticStimulus` aliases

## Follow-up work

Required follow-up:
- dynamic morphology section-type selection UI (#767)

Additional enhancements tracked separately:
- deterministic morphology-location blocks (#769)
- interactive morphology location selection in the morphology viewer (#770)
- exposing generated `compartment_sets.json` in Simulation input files (#768)

## Tests

Added coverage for:
- morphology-location materialization
- compartment-set export
- morphology path resolution
- legacy stimulus deserialization
- morphology-location validation

ilkilic committed:
refactor stimulus

ilkilic committed:
add compartment_set form

ilkilic committed:
backward compatiblity for stimuli

ilkilic committed:
lint fixes

ilkilic committed:
fix

ilkilic committed:
update unit-test

ilkilic committed:
refactor

ilkilic committed:
update bluecellulab version

ilkilic committed:
add validator

ilkilic committed:
update circuit simulation example

ilkilic committed:
lint fix

ilkilic committed:
cosmetic

ilkilic committed:
uv lock upgrade entitysdk

ilkilic committed:
refactoring

ilkilic committed:
WIP: materialize morphology-location targeting into compartment_set

ilkilic committed:
WIP: materialize morphology-location targeting into compartment_set

ilkilic committed:
Implement locations->compartment_set materialization pipeline

ilkilic committed:
minor fix

ilkilic committed:
update examples

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
lint fixes

ilkilic committed:
materialize stimulus locations into compartment sets

ilkilic committed:
Add missing json_schema_extra to scientific blocks

ilkilic committed:
clean notebook

ilkilic committed:
Generalize current clamp stimuli for morphology targeting

ilkilic committed:
Resolve morphology paths from SONATA components and manifests

ilkilic committed:
Materialize morphology locations into SONATA compartment sets

ilkilic committed:
fix

ilkilic committed:
cosmetic

ilkilic committed:
lint fix

ilkilic committed:
fix

ilkilic committed:
add notebook example

ilkilic committed:
rename examples

ilkilic committed:
add unit-test

ilkilic committed:
Support tuple section type sweeps for morphology locations

ilkilic committed:
Test morphology location materialization

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
Group morphology locations under targeting

ilkilic committed:
Hide section type selector pending UI support

ilkilic committed:
update default section types

ilkilic committed:
lint fix

ilkilic committed:
Merge branch 'main' into compartment-stimuli

james-isbister commented on obi_one/scientific/tasks/generate_simulations/materialize_locations.py:
```diff
@@ -0,0 +1,63 @@
+
+def materialize_locations_to_compartment_sets(
+    *,
+    form: Any,
```
I think we can use 'single_config' rather than form

ilkilic replied:
Done.

james-isbister commented on obi_one/scientific/tasks/generate_simulations/config/base.py:
```diff
@@ -50,12 +52,34 @@
 DEFAULT_TIMESTAMPS_NAME = "Default: Simulation Start (0 ms)"
 DEFAULT_DISTRIBUTION_NAME = "Default: Exp, scale 50 ms, 20 Hz"
 
+_STIMULUS_TYPE_ALIASES = {
```
It would be nice to change the names of these Python classes but the biggest issue is the duplication functionality in the frontend. If someone duplicates an old config in the frontend which has the old names, then it will be invalid in the frontend.

Two approaches we might take are:
- Renaming the old classes and creating aliases which are included in the stimulus union but are hidden (i.e. defining class ConstantCurrentClampSomaticStimulus(ConstantCurrentClampStimulus): pass
- Migrating the old configs

I think migrating the configs in entitycore would be the best longer term approach. In this PR I would suggest we revert to the old names with Somatic, and deal with this in a follow up, once everything is working in the frontend.

ilkilic replied:
Sounds good. I have reverted to the old names with Somatic.

james-isbister commented on obi_one/scientific/blocks/stimuli/stimulus.py:
```diff
@@ -119,6 +125,55 @@ class ContinuousStimulusWithoutTimestamps(BaseStimulus):
         },
     )
 
+    compartment_set: CompartmentSetReference | None = Field(
```
In the new_neuron_sets branch we have changed the REFERENCE_TYPE key in the schema of the ScanConfig block_dictionary to  represent in the schema that it holds multiple [REFERENCE_TYPES](https://github.com/openbraininstitute/obi-one/blob/517902ea48cf9998407f469201810663935dbc94/obi_one/scientific/tasks/generate_simulations/config/circuit.py#L81) in a list, note the plural.

The REFERENCE_TYPE in the schema of blocks has also changed to a list, plural [REFERENCE_TYPES](https://github.com/openbraininstitute/obi-one/blob/517902ea48cf9998407f469201810663935dbc94/obi_one/scientific/blocks/stimuli/stimulus.py#L118)

This means that a reference_type parameter of a block can now "hunt" for block_references of different types across multiple block dictionaries, and that each block_dictionary can serve different types of block reference.

This means that we can have a single dropdown for selecting the `target` of say a stimulus, which will show references from neuron_set and compartment_set block_dictionaries

I think we can rebase this PR to the new_neuron_sets branch to use this functionality. Although I'm not sure when the best time to merge will be as @chr-pok is also working on that

This removes the need for compartment_set and locations references here, and we can have a single `target`, although for now I would keep the name neuron_set for backwards compatibility, and we can think about migration in a separate PR once this is all working in the frontend.

ilkilic replied:
Thanks, agreed. I did not go in that direction yet because I wasn’t sure whether new_neuron_sets was stable enough to depend on already. Do you think it would make sense to rebase onto it now or would it be better to wait until that work is merged?

ilkilic replied:
Implemented after rebasing. Stimuli now use the existing `neuron_set` field as a single target selector, supporting neuron sets, compartment sets and morphology locations.

james-isbister commented on obi_one/scientific/blocks/morphology_locations/base.py:
```diff
@@ -6,23 +6,43 @@
 
 
 class MorphologyLocationsBlock(Block, abc.ABC):
     """Base class representing parameterized locations on morphology skeletons."""
```
I think this should take a neuron_set reference, and then locations should be calculated across all neurons in the neuron set

ilkilic replied:
Done. MorphologyLocationsBlock now has a neuron_set reference and location materialization uses that target.

james-isbister reviewed (commented):
Looking good @ilkilic Thanks and sorry for the delay

Here are my initial higher-level comments. I can look more at some of the details once we've decided on an approach for the existing comments

ilkilic committed:
Address compartment stimulus review comments

ilkilic committed:
Default morphology location targets for MEModel

ilkilic committed:
lint fixes

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
lint fix

ilkilic committed:
Scope morphology locations to neuron simulation configs

ilkilic committed:
add __init__.py

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
Revert unintended task endpoint exception handling change

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
fix: remove stale neuron set cast from compartment materialization

ilkilic committed:
refactor: unify continuous stimulus targeting

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
increase test coverage

james-isbister commented on obi_one/scientific/unions/unions_morphology_locations_ref.py:
```diff
@@ -0,0 +1,10 @@
+from typing import Any, ClassVar
```
Maybe the unions are references for MorphologyLocations could be put in the same file within the new unions_and_references directory to be consistent with the other types

ilkilic replied:
Done, I moved MorphologyLocationsReference into unions_morphology_locations.py so the union and reference live together, consistent with the other union/reference modules.

james-isbister commented on obi_one/scientific/blocks/morphology_locations/base.py:
```diff
@@ -6,23 +6,56 @@
+            SchemaKey.UI_ELEMENT: UIElement.INT_PARAMETER_SWEEP,
+        },
     )
+
```
What if a neuron doesn't have these section types? Maybe they would be fetched based on the neuron set in future, but don't need for now.

james-isbister replied:
Also I think the typing could be something like this:

> from typing import Literal
> SectionType = Literal["axon", "basal", "apical"]
> SectionTypes = tuple[SectionType, ...] | list[tuple[SectionType, ...]]
>
> section_types: SectionTypes = Field(
>     default=("axon", "basal", "apical"),
>     title="Section Types",
>     description=(
>         "Valid neurite section types to generate locations on: "
>         "axon, basal (basal dendrite), apical (apical dendrite). "
>         "Use a tuple for one selection, e.g. ('basal', 'apical'), "
>         "or a list of tuples for parameter scans."
>     ),
>     json_schema_extra={
>         SchemaKey.UI_HIDDEN: True,
>     },
> )

Or: 

> from typing import Literal
> SectionType = Literal[2, 3, 4]
> SectionTypes = tuple[SectionType, ...] | list[tuple[SectionType, ...]]
> 
> section_types: SectionTypes = Field(
>     default=(2, 3, 4),
>     title="Section Types",
>     description=(
>         "Valid neurite section types to generate locations on: "
>         "2: axon, 3: basal dendrite, 4: apical dendrite. "
>         "Use a tuple for one selection, e.g. (3, 4), "
>         "or a list of tuples for parameter scans."
>     ),
>     json_schema_extra={
>         SchemaKey.UI_HIDDEN: True,
>     },
> )

ilkilic replied:
Good point. I kept the MorphIO numeric values but restricted them to 2, 3, and 4. For morphologies missing a selected type, I think the follow-up feature/morphology-section-type-selection branch (https://github.com/openbraininstitute/obi-one/pull/822) should handle that by fetching available section types.

james-isbister commented on obi_one/scientific/blocks/stimuli/stimulus.py:
```diff
@@ -142,13 +177,20 @@ def config(
+                    NeuronSetPopulationType.POINT,
+                    NeuronSetPopulationType.NONVIRTUAL,
+                }
+            )
```
This check is maybe redundant now we have the typed population type specific neuron set references

james-isbister commented on obi_one/scientific/blocks/compartment_sets.py:
```diff
@@ -0,0 +1,165 @@
+class CompartmentSet(Block):
+    """SONATA-compatible compartment_set block."""
+
+    population: str = Field(
```
Here we can use the pattern of selecting a population within the BiophysicalPopulationNeuronSet, so we have a dropdown with only biophysical populations

james-isbister commented on obi_one/scientific/tasks/generate_simulations/task/task.py:
```diff
@@ -322,6 +365,22 @@ def _default_neuron_set_ref(self) -> ALL_NEURON_SETS_REFERENCE_UNION:
     NEURON SETS HAVE TYPES (BIOPHYSICAL, POINT, ETC.)
     """
 
+    @staticmethod
```
Does this need to be passed as an argument?

ilkilic replied:
Good point, removed the redundant circuit argument and uses self._circuit directly.

james-isbister replied:
Sorry, this is my mistake. The comment should have been higher up. I was referring to morphology_loader=self._load_morphology,

Is it possible that we would use a different loading function? Or could avoid passing the function here?

I think the passing the circuit in the function below as before was nice actually

ilkilic replied:
Ah sorry, yes, that makes sense. I removed the morphology_loader callback and now the materialization helper uses the passed circuit directly to load morphologies.

james-isbister commented on obi_one/scientific/tasks/generate_simulations/task/task.py:
```diff
@@ -428,6 +487,29 @@ def _resolve_neuron_sets_and_write_simulation_node_sets_file(self) -> None:
+            for cs_key, cs_block in self._materialized_compartment_sets.items():
+                if cs_key != cs_block.block_name:
+                    msg = "Materialized compartment set name mismatch."
+                    raise OBIONEError(msg)
```
If we create _materialized_compartment_sets, then would this error ever be triggered?

ilkilic replied:
Correct, this case should not happen in the normal task flow. I removed it.

james-isbister replied:
I still see the check on line 499

ilkilic replied:
Ah sorry, should be removed now.

james-isbister commented on obi_one/scientific/tasks/generate_simulations/task/task.py:
```diff
@@ -428,6 +487,29 @@ def _resolve_neuron_sets_and_write_simulation_node_sets_file(self) -> None:
+        # 4. If the config contains compartment_sets, collect them and write a
+        # compartment_sets.json file next to the node sets so stimuli referencing
+        # named compartment sets will resolve correctly.
+        if self._materialized_compartment_sets:
```
This doesn't really match the purpose of the function name "_resolve_neuron_sets_and_write_simulation_node_sets_file"

Maybe this could be combined with the code in materialize_location_targets with a name that includes the writing, if it makes sense?

ilkilic replied:
Good point. I moved the compartment-set writing into a separate `_write_materialized_compartment_sets_file`() method and call it explicitly after node sets are written.

james-isbister commented on obi_one/scientific/tasks/generate_simulations/materialize_locations.py:
```diff
@@ -0,0 +1,68 @@
+            raise ValueError(msg)
+
+        comp_set_name = f"{stimulus.block_name}__locations"
+
```
I'm not sure we need the block type CompartmentSet, and the CompartmentSetUnion, CompartmentSetReference etc.

The MorphologyLocation blocks could just have a function which resolves to the sonata entry. Wdyt?

ilkilic replied:
Good point, I agree. I refactored this so compartment sets are no longer public blocks/references. MorphologyLocations remains the user-facing API and materialization now produces the SONATA compartment_sets.json entry internally. If we later want users to define compartment sets directly we can add that as a separate advanced block with a clearer workflow.

james-isbister reviewed (commented):
Nice work! Here are my comments

ilkilic committed:
Move morphology locations reference into union module

ilkilic committed:
Remove redundant stimulus target population check

ilkilic committed:
Use biophysical population dropdown for compartment sets

ilkilic committed:
Simplify morphology loader callback

ilkilic committed:
Remove unsupported morphology target guard

ilkilic committed:
Separate compartment set file writing

ilkilic committed:
Keep compartment sets internal

ilkilic committed:
Constrain morphology section type selection

ilkilic committed:
Use circuit directly for morphology materialization

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
Merge branch 'main' into compartment-stimuli

james-isbister commented on examples/obi_one/scientific/tasks/generate_simulations/NEURON/circuit/run_circuit_simulation.ipynb:
Could just clear the notebook I think

ilkilic replied:
Done in the last commit

ilkilic committed:
Remove redundant morphology location target guard

ilkilic committed:
clear outputs in E_run_small_microcircuit/run_circuit_simulation.ipynb

james-isbister commented on obi_one/scientific/tasks/generate_simulations/materialize_locations.py:
```diff
@@ -0,0 +1,55 @@
+
+        locations_block = target_ref.block
+
+        neuron_set_ref = getattr(locations_block, "neuron_set", None)
```
The default is None (i.e. "Default: All Biophysical Neurons"), so this is very possible and should not raise an error

james-isbister replied:
This also brought to mind another issue. For large circuits, this is probably too computationally and/or memory expensive to be run on the service. Maybe this is where simulation generation would eventually need to be run on the launch system? Obviously not for this ticket.

Probably for now we should hide the morphology locations blocks if the circuit is larger than small microcircuit to begin? Using the usability feature? and a corresponding bool in the mapped circuit properties? Have you used that pattern before? I can explain otherwise?

ilkilic committed:
lint fix

james-isbister reviewed (commented):
Getting there I think. One thing about running campaign generation on the service just occurred to be though. 

I also replied to a small comment outside the review.

ilkilic committed:
Gate morphology location targets by circuit usability

james-isbister reviewed (dismissed):
LGTM!

ilkilic committed:
Upload materialized compartment sets with simulations

ilkilic committed:
Cover morphology location usability scale gate

ilkilic committed:
Use valid asset label for compartment sets

ilkilic committed:
update asset label

ilkilic committed:
Expect compartment sets asset label in upload test

ilkilic committed:
Resolve relative morphology file paths

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
uv lock

ilkilic committed:
bump entitysdk

ilkilic committed:
uv lock

ilkilic committed:
Merge branch 'main' into compartment-stimuli

james-isbister commented on obi_one/scientific/tasks/generate_simulations/config/neuron/neuron_me_model.py:
```diff
@@ -88,12 +91,14 @@ class Initialize(NeuronSimulationScanConfig.Initialize):
         ],
         SchemaKey.DEFAULT_BLOCK_REFERENCE_LABELS: {
             TimestampsReference.__name__: DEFAULT_TIMESTAMPS_NAME,
+            MorphologyLocationsReference.__name__: "Default: None",
```
`Default: No locations` is maybe better?

james-isbister commented on obi_one/scientific/tasks/generate_simulations/config/base.py:
```diff
@@ -52,6 +52,7 @@ class BlockGroup(StrEnum):
 
     SETUP_BLOCK_GROUP = "Setup"
     STIMULI_RECORDINGS_BLOCK_GROUP = "Stimuli & Recordings"
+    TARGETING_BLOCK_GROUP = "Targeting"
```
I think we can group Neuron Sets and Morphology Locations into a single BlockGroup but I'm not sure "Circuit Components" or "Targeting". Maybe we can think of a different term in a follow up PR

ilkilic replied:
I tried `Targets` in the follow-up PR. It groups Neuron Sets and Morphology Locations together, wdyt?

james-isbister reviewed (dismissed):
LGTM! And works well in the frontend!

I think we can merge the other branch into this one, and then merge this to main after the staging freeze if there is one.

I would make the following textual changes in a separate PR so we can consider the changes clearly:
- Change: MorphologyLocationsReference.__name__: "Default: No Locations"
- Find a better BlockGroup name for grouping Neuron Sets and Morphology Locations
- Improve the block and parameter descriptions of the Morphology Locations to explain how each of them work. I hadn't looked at the Morphology Locations for a while, so I was unclear how the different combinations of clusters and locations would lead to different outcomes (and what each of those parameters meant exactly)
- Add units for some of the distance based parameters of the MorphoogyLocations Blocks
- Add title class variables to the MorphologyLocations blocks with spaces. Currently the class name without spaces is displayed in the frontend

Would be good to also test the performance in staging once its live, and think about expanding to larger circuits.

ilkilic committed:
Add dynamic morphology section-type discovery and schema support (#822)

* Support section type options for small circuits

* Restrict morphology locations to dendrite section types

ilkilic committed:
Validate morphology property enum values

ilkilic committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
lint fix

james-isbister commented on obi_one/scientific/blocks/morphology_locations/clustered.py:
I think the check `if self.n_clusters < 1:` is no longer needed now `n_clusters: PositiveInt`

ilkilic replied:
I updated it in https://github.com/openbraininstitute/obi-one/pull/932

james-isbister commented on obi_one/scientific/blocks/morphology_locations/clustered.py:
This still raises an unknown error in the UI. Not sure why the message isn't displayed

<img width="888" height="324" alt="Image" src="https://github.com/user-attachments/assets/3fce12a4-78fe-4f2e-bfce-c782a6f7eb03" />

ilkilic replied:
I had a quick look in core-web-app and I wonder if the issue might be there. It looks like the coordinate-count handler reads message.detail while the backend returns the validation message in details[0].msg.

@bilalesi would you mind having a quick look? Do you think the coordinate-count handler should use details[0].msg here as well? Thank you!

james-isbister reviewed (commented):
Looking good in the UI. Just one final thing to look at + a minor change

james-isbister committed:
Merge branch 'main' into compartment-stimuli

ilkilic committed:
Polish morphology location schema various (#932)

- Rename the default morphology-location reference label to `Default: No Locations`
- Group NEURON neuron sets and morphology locations under `Targets`
- Add readable titles for morphology-location block types
- Improve descriptions for morphology-location parameters and block dictionaries
- Add micrometer units to distance-based morphology-location parameters

james-isbister committed:
Merge branch 'compartment-stimuli' of https://github.com/openbraininstitute/obi-one into compartment-stimuli

james-isbister committed:
Formatting
